README.md
# LAWS

A [Lua](http://www.lua.org/) library for interacting with [AWS (Amazon Web Services)](https://aws.amazon.com/).

## Dependencies

  - [luaossl](http://25thandclement.com/~william/projects/luaossl.html)


## Modules

### `aws.v4`

Implements the [AWS Signature v4](http://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html) algorithm.
This can be used to sign requests to a variety of Amazon services.

#### `aws.v4.canonicalise_path`

```lua
CanonicalURI = canonicalise_path(path)
```

Canonicalises the given string argument according to the v4 signature rules for paths:

  - Removes redundant path components, e.g. `..` and `.`
  - Ensures (only) the correct characters are escaped


#### `aws.v4.canonicalise_query_string`

```lua
CanonicalQueryString = canonicalise_query_string(query)
```

Canonicalises the given string argument according to the v4 signature rules for query strings:

  - Ensures (only) the correct characters are escaped
  - Query arguments are sorted


#### `aws.v4.derive_signing_key`

```lua
kSigning = derive_signing_key(kSecret, Date, Region, Service)
```

Derives the signing key as specified in http://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html
Uses a SHA256 HMAC.


#### `aws.v4.prepare_request`

```lua
request, extra = prepare_request(tbl)
```

Prepares a signed HTTP(S) request.

Currently does not support Authentication parameters in the query string.


The fields of `tbl` are:

Field                   | Type                          | Description
------------------------|-------------------------------|-----------------------
`.Domain`               | `string\|nil`                 | The domain to hit, defaults to `"amazonaws.com"`
`.Region`               | `string`                      | The Amazon region. e.g. `"us-east-1"`
`.Service`              | `string`                      | The Amazon service. e.g. `"kinesis"`
`.method`               | `string`                      | e.g. `"GET"`
`.CanonicalURI`         | `string\|nil`                 | A pre-canonicalised path. If not provided `.path` will be canonicalised
`.path`                 | `string\|nil`                 | The path of your request. Not used if `.CanonicalURI` is specified. If not provided or empty, defaults to `"/"`
`.CanonicalQueryString` | `string\|nil`                 | A pre-canonicalised query-string. If not provided `.query` will be canonicalised
`.query`                | `string\|nil`                 | The query string for your request. Not used if `.CanonicalQueryString` is specified. Not required.
`.headers`              | `{string:string\|false}\|nil` | Table of string key/value pairs to use as signed headers in the request. A value of `false` prevents the header from being automatically filled in. Keys will be normalised to Title-Case; behaviour on duplicates is undefined.
`.body`                 | `string\|nil`                 | Request body
`.AccessKey`            | `string`                      | Your Amazon access key.
`.SigningKey`           | `string\|nil`                 | A pre-derived signing key, if one is not provided, it will be derived from `.SecretKey`
`.SecretKey`            | `string\|nil`                 | Your Amazon secret key, only required if you do not specify `.SigningKey`
`.timestamp`            | `number\|nil`                 | When request should be signed for as a unix timestamp. defaults to `os.time()` i.e. now
`.tls`                  | `boolean\|nil`                | A truthy value will use HTTPS, false indicates HTTP. defaults to `true`
`.port`                 | `number\|nil`                 | TCP port for request. defaults to `443` for HTTPS or `80` for HTTP

The returned `request` object has fields:

Field       | Type              | Description
------------|-------------------|-----------------------------------------------
`.url`      | `string`          | URL representing the Host,
`.host`     | `string`          | Host to connect to. e.g. `"kinesis.us-east-1.amazonaws.com"`
`.port`     | `number`          | TCP port to connect to. e.g. `443`
`.tls`      | `boolean`         | Should SSL/TLS be used to connect? e.g. `true`
`.method`   | `string`          | e.g. `"GET"`
`.target`   | `string`          | The target (the part between method and `HTTP/`) ready to be used in a HTTP request. i.e. path and query string
`.headers`  | `{string:string}` | Table of headers
`.body`     | `string\|nil`      | `.body` as passed in

The returned `extra` object has intermediary values calculated by the AWS v4 signing algorithm.

Field                   | Type
------------------------|----------
`.CanonicalURI`         | `string`
`.CanonicalQueryString` | `string`
`.SignedHeaders`        | `string`
`.CanonicalHeaders`     | `string`
`.CanonicalRequest`     | `string`
`.StringToSign`         | `string`
`.SigningKey`           | `string`
`.Signature`            | `string`
`.Authorization`        | `string`


If not provided, certain headers will be added.
To prevent a header from getting added, provide it in `.headers` with a boolean value of `false`

  - `Host`
  - `X-Amz-Date`
  - `Authorization`


##### Example

This example prepares a HTTP request to Amazon Kinesis to [ListStreams](http://docs.aws.amazon.com/kinesis/latest/APIReference/API_ListStreams.html)

```lua
local prepare_request = require "aws.v4".prepare_request
local request, extra = prepare_request {
	Region = "us-east-1";
	Service = "kinesis";
	method = "POST";
	headers = {
		["X-Amz-Target"] = "Kinesis_20131202.ListStreams";
		["Content-Type"] = "application/x-amz-json-1.1";
	};
	body = '{}';
	AccessKey = "AKIDEXAMPLE";
	SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
}
```


NEWS
0.1 - 2018-08-24

  - Tag of code after long-dormant state


spec/v4_spec.lua
local awsv4 = require "aws.v4"

describe("Pass AWSv4 test suite", function()
	-- The test suite was obtained from https://docs.aws.amazon.com/general/latest/gr/samples/aws-sig-v4-test-suite.zip
	-- sha1sum of the test suite .zip is c66ea06884dad4348558e93380c8bf91e8551a6f
	-- Information about it can be found at http://docs.aws.amazon.com/general/latest/gr/signature-v4-test-suite.html
	local function read_file(path)
		local fd, err, code = io.open(path, "rb")
		if fd == nil then
			if code == 2 then -- not found
				return nil
			else
				error(err)
			end
		end
		local contents = assert(fd:read"*a")
		fd:close()
		return contents
	end

	local dir = "./spec/aws-sig-v4-test-suite/"
	for _, test_name in ipairs {
		"get-header-key-duplicate";
		"get-header-value-multiline";
		"get-header-value-order";
		"get-header-value-trim";
		"get-unreserved";
		"get-utf8";
		"get-vanilla";
		"get-vanilla-empty-query-key";
		"get-vanilla-query";
		"get-vanilla-query-order-key";
		"get-vanilla-query-order-key-case";
		"get-vanilla-query-order-value";
		"get-vanilla-query-unreserved";
		"get-vanilla-utf8-query";
		"normalize-path/get-relative";
		"normalize-path/get-relative-relative";
		"normalize-path/get-slash";
		"normalize-path/get-slash-dot-slash";
		"normalize-path/get-slashes";
		"normalize-path/get-slash-pointless-dot";
		"normalize-path/get-space";
		"post-header-key-case";
		"post-header-key-sort";
		"post-header-value-case";
		"post-sts-token/post-sts-header-after";
		"post-sts-token/post-sts-header-before";
		"post-vanilla";
		"post-vanilla-empty-query-value";
		"post-vanilla-query";
		"post-x-www-form-urlencoded";
		"post-x-www-form-urlencoded-parameters";
	} do
		local filename = test_name:match("[^/]+$")
		local req = read_file(dir..test_name.."/"..filename..".req")
		;(req and it or pending)("passes test #" .. test_name:gsub("%-", "_"), function()
			local creq = read_file(dir..test_name.."/"..filename..".creq")
			local sts = read_file(dir..test_name.."/"..filename..".sts")
			local authz = read_file(dir..test_name.."/"..filename..".authz")
			-- local sreq = read_file(dir..test_name..".sreq")

			local method, target, pos = req:match("^(%S+) (.-) HTTP/1.[01]()")
			local path, query = target:match("([^%?]*)%??(.*)")
			path = path or target
			local str_headers, body = req:match("^\n(.-\n)\n(.*)", pos)
			str_headers = str_headers or req:sub(pos+1) .. "\n"
			local headers = {
				["X-Amz-Date"] = false; -- test suite uses normal date header instead
			}
			for k, v in str_headers:gmatch("([^%s:]+): *(.-)\n%f[%z%w]") do
				v = v:gsub("\n +", ",")
				local old_v = headers[k:lower()]
				if old_v then
					v = old_v .. "," .. v
				end
				headers[k:lower()] = v
			end
			if body == "" then body = nil end
			local http_req, interim = awsv4.prepare_request {
				Region = "us-east-1";
				Service = "service";
				AccessKey = "AKIDEXAMPLE";
				SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
				method = method;
				path = path;
				query = query;
				headers = headers;
				body = body;
				timestamp = 1440938160; -- Timestamp used by all tests
			}
			assert.truthy(http_req)
			assert.same(creq, interim.CanonicalRequest)
			assert.same(sts, interim.StringToSign)
			assert.same(authz, interim.Authorization)
		end)
	end
end)

describe("Path canonicalisation is correct", function()
	it("Handles . and .. correctly", function()
		assert.same("/", awsv4.canonicalise_path "/")
		assert.same("/", awsv4.canonicalise_path "/.")
		assert.same("/", awsv4.canonicalise_path "/./foo/../")
		assert.same("/bar", awsv4.canonicalise_path "/foo/../foo/./../bar")
		assert.same("/..foo", awsv4.canonicalise_path "/..foo")
		assert.same("/bar/.foo", awsv4.canonicalise_path "/./bar/.foo")
	end)
	it("Can't get above top dir", function()
		assert.same("/foo", awsv4.canonicalise_path "/../foo")
	end)
	it("Escapes correctly", function()
		assert.same("/foo", awsv4.canonicalise_path "/%66oo")
		-- for aws, space must be %20, not +
		assert.same("/%20", awsv4.canonicalise_path "/ ")
	end)
end)

describe("port is handled correctly", function()
	it("is appended to Host", function()
		assert.same("host.us-east-1.amazonaws.com", (awsv4.prepare_request {
			Region = "us-east-1";
			Service = "host";
			AccessKey = "AKIDEXAMPLE";
			SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
			method = "GET";
		}).headers.Host)
		assert.same("host.us-east-1.amazonaws.com", (awsv4.prepare_request {
			Region = "us-east-1";
			Service = "host";
			AccessKey = "AKIDEXAMPLE";
			SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
			method = "GET";
			port = 443;
		}).headers.Host)
		assert.same("host.us-east-1.amazonaws.com", (awsv4.prepare_request {
			Region = "us-east-1";
			Service = "host";
			AccessKey = "AKIDEXAMPLE";
			SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
			method = "GET";
			port = 80;
			tls = false
		}).headers.Host)
		assert.same("host.us-east-1.amazonaws.com:1234", (awsv4.prepare_request {
			Region = "us-east-1";
			Service = "host";
			AccessKey = "AKIDEXAMPLE";
			SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
			method = "GET";
			port = 1234;
		}).headers.Host)
		assert.same("host.us-east-1.amazonaws.com:8080", (awsv4.prepare_request {
			Region = "us-east-1";
			Service = "host";
			AccessKey = "AKIDEXAMPLE";
			SecretKey = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY";
			method = "GET";
			port = 8080;
			tls = false
		}).headers.Host)
	end)
end)


aws/v4.lua
-- Performs AWSv4 Signing
-- http://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html

local openssl_hmac_new = require "openssl.hmac".new
local openssl_digest_new = require "openssl.digest".new

local Algorithm = "AWS4-HMAC-SHA256"
local function HMAC(key, msg)
	return openssl_hmac_new(key, "sha256"):final(msg)
end
local function Hash(str)
	return openssl_digest_new("sha256"):final(str)
end
local HexEncode do -- From prosody's util.hex
	local char_to_hex = {};
	for i = 0, 255 do
		local char = string.char(i)
		local hex = string.format("%02x", i)
		char_to_hex[char] = hex
	end
	function HexEncode(str)
		return (str:gsub(".", char_to_hex))
	end
end

local function percent_encode(char)
	return string.format("%%%02X", string.byte(char))
end

local urldecode do
	local function urldecode_helper(c)
		return string.char(tonumber(c, 16))
	end
	function urldecode(str)
		return (str:gsub("%%(%x%x)", urldecode_helper))
	end
end

-- Based on Trim 12 from http://lua-users.org/wiki/StringTrim
local function Trimall(s)
	local from = s:match"^%s*()"
	return from > #s and "" or s:match(".*%S", from):gsub("  *", " ")
end

local function canonicalise_path(path)
	local segments = {}
	for segment in path:gmatch("/([^/]*)") do
		if segment == "" or segment == "." then -- luacheck: ignore 542
			-- nothing
		elseif segment == ".." then
			-- intentionally discards components at top level
			segments[#segments] = nil
		else
			segments[#segments+1] = urldecode(segment):gsub("[^%w%-%._~]", percent_encode)
		end
	end
	local len = #segments
	if len == 0 then return "/" end
	-- If there was a slash on the end, keep it there.
	if path:sub(-1, -1) == "/" then
		len = len + 1
		segments[len] = ""
	end
	segments[0] = ""
	segments = table.concat(segments, "/", 0, len)
	return segments
end

local function canonicalise_query_string(query)
	local q = {}
	for key, val in query:gmatch("([^&=]+)=?([^&]*)") do
		key = urldecode(key):gsub("[^%w%-%._~]", percent_encode)
		val = urldecode(val):gsub("[^%w%-%._~]", percent_encode)
		q[#q+1] = key .. "=" .. val
	end
	table.sort(q)
	return table.concat(q, "&")
end

local function derive_signing_key(kSecret, Date, Region, Service)
	local kDate = HMAC("AWS4" .. kSecret, Date)
	local kRegion = HMAC(kDate, Region)
	local kService = HMAC(kRegion, Service)
	local kSigning = HMAC(kService, "aws4_request")
	return kSigning
end

local function prepare_awsv4_request(tbl)
	local Domain = tbl.Domain or "amazonaws.com"
	assert(type(Domain) == "string", "bad field 'Domain' (string or nil expected)")
	local Region = tbl.Region
	assert(type(Region) == "string", "bad field 'Region' (string expected)")
	local Service = tbl.Service
	assert(type(Service) == "string", "bad field 'Service' (string expected)")
	local HTTPRequestMethod = tbl.method
	assert(type(HTTPRequestMethod) == "string", "bad field 'method' (string expected)")
	local CanonicalURI = tbl.CanonicalURI
	local path = tbl.path
	if CanonicalURI == nil and path ~= nil then
		assert(type(path) == "string", "bad field 'path' (string or nil expected)")
		CanonicalURI = canonicalise_path(path)
	elseif CanonicalURI == nil or CanonicalURI == "" then
		CanonicalURI = "/"
	end
	assert(type(CanonicalURI) == "string", "bad field 'CanonicalURI' (string or nil expected)")
	local CanonicalQueryString = tbl.CanonicalQueryString
	local query = tbl.query
	if CanonicalQueryString == nil and query ~= nil then
		assert(type(query) == "string", "bad field 'query' (string or nil expected)")
		CanonicalQueryString = canonicalise_query_string(query)
	end
	assert(type(CanonicalQueryString) == "string" or CanonicalQueryString == nil, "bad field 'CanonicalQueryString' (string or nil expected)")
	local req_headers = tbl.headers or {}
	assert(type(req_headers) == "table", "bad field 'headers' (table or nil expected)")
	local RequestPayload = tbl.body
	assert(type(RequestPayload) == "string" or RequestPayload == nil, "bad field 'body' (string or nil expected)")
	local AccessKey = tbl.AccessKey
	assert(type(Region) == "string", "bad field 'AccessKey' (string expected)")
	local SigningKey = tbl.SigningKey
	assert(type(SigningKey) == "string" or SigningKey == nil, "bad field 'SigningKey' (string or nil expected)")
	local SecretKey
	if SigningKey == nil then
		SecretKey = tbl.SecretKey
		if SecretKey == nil then
			assert(SecretKey, "either 'SigningKey' or 'SecretKey' must be provided")
		end
		assert(type(SecretKey) == "string", "bad field 'SecretKey' (string expected)")
	end
	local timestamp = tbl.timestamp or os.time()
	assert(type(timestamp) == "number", "bad field 'timestamp' (number or nil expected)")
	local tls = tbl.tls
	if tls == nil then tls = true end
	local port = tbl.port or (tls and 443 or 80)
	assert(type(port) == "number", "bad field 'port' (string or nil expected)")

	local RequestDate = os.date("!%Y%m%dT%H%M%SZ", timestamp)
	local Date = os.date("!%Y%m%d", timestamp)

	local host = Service .. "." .. Region .. "." .. Domain
	local host_header do -- If the "standard" port is not in use, the port should be added to the Host header
		local with_port
		if tls then
			with_port = port ~= 443
		else
			with_port = port ~= 80
		end
		if with_port then
			host_header = string.format("%s:%d", host, port)
		else
			host_header = host
		end
	end

	local headers = {
		["X-Amz-Date"] = RequestDate;
		Host = host_header;
	}
	local add_auth_header = true
	for k, v in pairs(req_headers) do
		assert(type(k) == "string", "bad field 'headers' (only string keys allowed)")
		k = k:gsub("%f[^%z-]%w", string.upper) -- convert to standard header title case
		if k == "Authorization" then
			add_auth_header = false
		elseif v == false then -- don't allow a default value for this header
			v = nil
		end
		headers[k] = v
	end

	-- Task 1: Create a Canonical Request For Signature Version 4
	-- http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
	local CanonicalHeaders, SignedHeaders do
		-- We structure this code in a way so that we only have to sort once.
		CanonicalHeaders, SignedHeaders = {}, {}
		local i = 0
		for name, value in pairs(headers) do
			if value then -- ignore headers with 'false', they are used to override defaults
				i = i + 1
				local name_lower = name:lower()
				SignedHeaders[i] = name_lower
				assert(CanonicalHeaders[name_lower] == nil, "header collision")
				CanonicalHeaders[name_lower] = Trimall(value)
			end
		end
		table.sort(SignedHeaders)
		for j=1, i do
			local name = SignedHeaders[j]
			local value = CanonicalHeaders[name]
			CanonicalHeaders[j] = name .. ":" .. value .. "\n"
		end
		SignedHeaders = table.concat(SignedHeaders, ";", 1, i)
		CanonicalHeaders = table.concat(CanonicalHeaders, nil, 1, i)
	end
	local CanonicalRequest =
		HTTPRequestMethod .. '\n' ..
		CanonicalURI .. '\n' ..
		(CanonicalQueryString or "") .. '\n' ..
		CanonicalHeaders .. '\n' ..
		SignedHeaders .. '\n' ..
		HexEncode(Hash(RequestPayload or ""))
	local HashedCanonicalRequest = HexEncode(Hash(CanonicalRequest))
	-- Task 2: Create a String to Sign for Signature Version 4
	-- http://docs.aws.amazon.com/general/latest/gr/sigv4-create-string-to-sign.html
	local CredentialScope = Date .. "/" .. Region .. "/" .. Service .. "/aws4_request"
	local StringToSign =
		Algorithm .. '\n' ..
		RequestDate .. '\n' ..
		CredentialScope .. '\n' ..
		HashedCanonicalRequest
	-- Task 3: Calculate the AWS Signature Version 4
	-- http://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html
	if SigningKey == nil then
		SigningKey = derive_signing_key(SecretKey, Date, Region, Service)
	end
	local Signature = HexEncode(HMAC(SigningKey, StringToSign))
	-- Task 4: Add the Signing Information to the Request
	-- http://docs.aws.amazon.com/general/latest/gr/sigv4-add-signature-to-request.html
	local Authorization = Algorithm
		.. " Credential=" .. AccessKey .."/" .. CredentialScope
		.. ", SignedHeaders=" .. SignedHeaders
		.. ", Signature=" .. Signature
	if add_auth_header then
		headers.Authorization = Authorization
	end

	local target = path or CanonicalURI
	if query or CanonicalQueryString then
		target = target .. "?" .. (query or CanonicalQueryString)
	end
	local scheme = tls and "https" or "http"
	local url = scheme .. "://" .. host_header .. target

	return {
		url = url;
		host = host;
		port = port;
		tls = tls;
		method = HTTPRequestMethod;
		target = target;
		headers = headers;
		body = RequestPayload;
	}, {
		CanonicalURI = CanonicalURI;
		CanonicalQueryString = CanonicalQueryString;
		SignedHeaders = SignedHeaders;
		CanonicalHeaders = CanonicalHeaders;
		CanonicalRequest = CanonicalRequest;
		StringToSign = StringToSign;
		SigningKey = SigningKey;
		Signature = Signature;
		Authorization = Authorization;
	}
end

return {
	canonicalise_path = canonicalise_path;
	canonicalise_query_string = canonicalise_query_string;
	derive_signing_key = derive_signing_key;
	prepare_request = prepare_awsv4_request;
}
